File: LeetCode/Medium/3259. Maximum Energy Boost From Two Drinks.cpp
// Link: https://leetcode.com/problems/maximum-energy-boost-from-two-drinks/
using ll=long long;
const int N=1e5+5;
class Solution {
public:

ll n;
vector<int>a,b;
ll dp[N][2];
ll fun(ll i,bool sw) {
    if(i==n)return 0;
    ll& ret=dp[i][sw];
    if(~ret)return ret;
    ret=0;
    if(sw==0) {
        ret=max(ret,fun(i+1,sw)+a[i]);
        ret=max(ret,fun(i+1,sw^1));
    }
    else {
        ret=max(ret,fun(i+1,sw)+b[i]);
        ret=max(ret,fun(i+1,sw^1));
    }
    return ret;
}
long long maxEnergyBoost(vector<int>&eA,vector<int>&eB) {
        
    for(int i=0;i<N;i++) {
        dp[i][0]=-1;
        dp[i][1]=-1;
    }
    a=eA;
    b=eB;
    n=a.size();
    return max(fun(0,0),fun(0,1));

    
    }
};

File: LeetCode/Medium/3259. Maximum Energy Boost From Two Drinks.py
# Link: https://leetcode.com/problems/maximum-energy-boost-from-two-drinks/
class Solution:

    def __init__(self):
        self.dp=[[]]
        self.a=[]
        self.b=[]
        self.n=0
        
    def fun(self, i , sw : int)-> int:
        if i==self.n:
            return 0
        if self.dp[i][sw]!=-1:
            return self.dp[i][sw]
        if sw==0:
            self.dp[i][sw]=max(self.dp[i][sw],self.fun(i+1,sw)+self.a[i])
            self.dp[i][sw]=max(self.dp[i][sw],self.fun(i+1,sw^1))
        else:
            self.dp[i][sw]=max(self.dp[i][sw],self.fun(i+1,sw)+self.b[i])
            self.dp[i][sw]=max(self.dp[i][sw],self.fun(i+1,sw^1))
        return self.dp[i][sw]
    def maxEnergyBoost(self, ea: List[int], eb: List[int]) -> int:
        self.a=ea
        self.b=eb
        self.n=len(self.a)
        self.dp=[ [-1 for _ in range(2)] for _ in range(self.n+5)]
        return max(self.fun(0,0),self.fun(0,1))

